### ISSUE TITLE
add example code and docs to demonstrate matching data: and tel: URLs

### ISSUE THREAD

**Issue by GoogleCodeExporter** [OWASP/java-html-sanitizer#21]
```
There is no existing documentation on how to allow data: and tel: URLs.  Add an 
example that demonstrates how HtmlPolicyBuilder.allowUrlProcotols can be used 
with data: and tel:.

```

Original issue reported on code.google.com by `mikesamuel@gmail.com` on 21 Jan 2014 at 4:04

**Comment by jmanico** [comment]
Bump to revisit. Mike, perhaps just give me one complex example and I'll use it to flesh out the docs.

**Comment by pickle-weasle** [comment]
Can't this be achieved with just 
`this.policy = this.policy.allowUrlProtocols("tel","ftp");`
In the [readme ](https://github.com/OWASP/java-html-sanitizer)there is the following:
`PolicyFactory policy = new HtmlPolicyBuilder()
    .allowElements("a")
    **.allowUrlProtocols("https")**
    .allowAttributes("href").onElements("a")
    .requireRelNofollowOnLinks()
    .toFactory();
String safeHTML = policy.sanitize(untrustedHTML);`
It's hardly much of a stretch to figure how to change that for other protocols.
Could this be closed?

**Comment by DavesMan** [comment]
> Can't this be achieved with just
> `this.policy = this.policy.allowUrlProtocols("tel","data");`
> In the [readme ](https://github.com/OWASP/java-html-sanitizer)there is the following:

No. I have allowUrlProtocols("mailto", "javascript", "tel", "https") and this allows https and mailto. But tel & javascript gets stripped.

**Comment by mikesamuel** [comment]
I put together https://github.com/OWASP/java-html-sanitizer/pull/126 to try to fill this gap.  It integrates a [URL classifier engine](https://github.com/OWASP/url-classifier) that lets you express predicates over URLs like

```java
  /** We define a classifier with a declarative syntax */
  static final UrlClassifier CLASSIFIER = UrlClassifiers.builder()
      // We want to allow HTTP and HTTPS for this example
      .scheme(BuiltinScheme.HTTP, BuiltinScheme.HTTPS)
      .authority(
          AuthorityClassifiers.builder()
          // We whitelist some subdomains of hosts we trust.
          .host("**.example.com", "**.example.net")
          .build())
      // We allow access to .html files
      .pathGlob("**.html")
      .build();
```

I haven't integrated it because I don't know if anyone needs that level of flexibility.

It does have better support for data and tel URLs.

https://github.com/OWASP/url-classifier/blob/8715ef5f8d5623b8023b6c59b931621431188f21/src/main/java/org/owasp/url/BuiltinScheme.java#L230-L241

If that looks useful, I can move that in.

**Comment by DavesMan** [comment]
Seems it could save a lot of "custom" work. No I also do some substring and regex checks of URLs using apply method of ElementPolicy.

**Comment by jmiserez** [comment]
I'm uncertain on whether this would actually solve all issues with `tel:` or other URIs.

It's currently possible to allow the `tel:` protocol using `.allowUrlProtocols("tel:")`, and that works. But AFAIK it's not possible to prevent `tel:` URIs from going through the normalization and escaping that `FilterUrlByProtocolAttributePolicy#normalizeUri` does.

Due to the way [RFC 3966 - 3. URI syntax](https://datatracker.ietf.org/doc/html/rfc3966#section-3) is written, at least escaping the leading `+` character is incorrect and actually leads to unusable tel: URIs:

>  The syntax definition follows [RFC 2396](https://datatracker.ietf.org/doc/html/rfc2396), indicating the
>  actual characters contained in the URI.  If the reserved characters
>  "+", ";", "=", and "?" are used as delimiters between components of
>  the "tel" URI, they MUST NOT be percent encoded.  These characters
>  MUST be percent encoded if they appear in tel URI parameter values.

Although the URIs do work in some browsers, they don't work in other applications (several email clients) due to this.

I guess my question is if this change already allows implements protocol-dependent handling of URIs or if it only allows more fine-grained restrictions on top of the existing processing, i.e. will it bypass `FilterUrlByProtocolAttributePolicy#normalizeUri` or not?

**Comment by jmiserez** [comment]
Just noticed that org.owasp.html.Encoding#encodeHtmlAttribOnto is responsible in the case of +, not FilterUrlByProtocolAttributePolicy#normalizeUri. 

I've edited that above, but the question remains the same.

**Comment by mikesamuel** [comment]
This issue is about docs and examples.
If you have a problem with mis-encoding of tel: URLs could you open a new issue with an example of a tel: URL which is mis-encoded.

**Comment by jmiserez** [comment]
Not sure about the hostility. This issue is clearly about more than just docs and examples, given the above PR.

While I _do_ have a problem with encoding tel: URIs, that's not what I'm asking.

What I _am_ asking is whether you had given any thought to protocol-dependent escaping of characters in the above PR. At the moment it looks like that is not the case. That affects more protocols than just tel:, especially new ones.

I'm happy to move this question/feedback over to https://github.com/OWASP/java-html-sanitizer/issues/22, but this issue seemed more current.

**Comment by jmanico** [comment]
There is no hostility in this thread that I see. We mean well.

**Comment by mikesamuel** [comment]
@jmiserez 

I'm trying to fit my pro-bono work, including this project, in between a lot of other stuff.  Sorry if I was abrupt, but I have no hostility towards you.

https://greatergood.berkeley.edu/article/item/six_tips_for_reading_emotions_in_text_messages

> Keep in mind that texts are a difficult medium for communicating emotion. We have no facial expressions, or tone of voice, or conversation to give us more information.
>
> If the text doesn’t say, “I’m angry,” then don’t assume that the texter is angry.

I have not read the `tel:` RFC recently.  It would help out if you could provide an example of what you want.  Yes, scope creep in issues happens, and it's happened in this thread.  It would help me out again though if, something with a concrete success case could happen in a thread that's narrowly tailored, or in the PR comments.

If issue #22 is where you want to discuss it that's great.  Recency doesn't matter to me.

> What I am asking is whether you had given any thought to protocol-dependent escaping of characters in the above PR.

That PR includes `TEL` as a *BuiltinScheme*:

https://github.com/OWASP/url-classifier/blob/8715ef5f8d5623b8023b6c59b931621431188f21/src/main/java/org/owasp/url/BuiltinScheme.java#L225-L228

*Scheme*'s control how raw parts are composed into a string.

https://github.com/OWASP/url-classifier/blob/8715ef5f8d5623b8023b6c59b931621431188f21/src/main/java/org/owasp/url/Scheme.java#L381-L392

I don't know if I got that correct, because I've not seen a failing testcase.  https://github.com/OWASP/url-classifier/blob/master/src/test/java/org/owasp/url/UrlValueTest.java has tests for encoding of percents so that, e.g. `javascript:alert(9 % 2)` doesn't have `%` reencoded to `%22` but it doesn't have many `tel:` examples.

**Comment by jmiserez** [comment]
I seem to have wildly misinterpreted your response and I'm sorry for the snide remark. I appreciate all the work and effort you put into this project and don't want to waste your time with nonsense. I must admit my initial comment concerning the PR wasn't too clear either.

Thank you for your detailed response, the part concerning javascript escaping is especially relevant and helpful in this context.

I really don't want to take up any more of your time, so I'll try out the test cases and the classifier library in more detail and comment on the PR itself or https://github.com/OWASP/java-html-sanitizer/issues/22 with a test case once I have something more tangible.

**Comment by jmanico** [comment]
Closing this in an old-issue triage pass — it has been open since 2013 and the substance has been answered along the way.

On the confusion that came up in the thread: `allowUrlProtocols("tel")` alone does not make a `tel:` link survive, which is what tripped up @DavesMan. Protocols and attributes are separate gates, and you need both, plus the element:

```java
new HtmlPolicyBuilder()
    .allowElements("a")
    .allowUrlProtocols("https", "mailto", "tel")
    .allowAttributes("href").onElements("a")
    .toFactory();
```

`javascript:` is the exception and stays dropped whatever you pass to `allowUrlProtocols`, by design.

For `data:` URLs restricted to images, @cnsgithub's working version is in #22 and is the pattern to copy.

The more ambitious idea here — @mikesamuel's #126, integrating the url-classifier engine for declarative predicates over scheme, host and path — never landed and has been dormant for years. Leaving this issue open does not make that any more likely; if someone wants to revive it, #126 is the thing to reopen, not this.

I have not folded these snippets into the README. If the docs are what you actually want, that is worth a fresh, narrow issue ("document protocol + attribute interaction for tel:/data: in the README") that can be closed by a concrete PR.
